File: backend/payment-service/src/main/java/lk/nibm/paymentservice/service/PaymentService.java
package lk.nibm.paymentservice.service;

import lk.nibm.common.dto.PaymentRequest;
import lk.nibm.common.dto.PaymentResponse;
import lk.nibm.paymentservice.model.Payment;
import lk.nibm.paymentservice.repository.PaymentRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.math.BigDecimal;
import java.util.List;
import java.util.Optional;
import java.util.UUID;

@Service
public class PaymentService {

    @Autowired
    private PaymentRepository paymentRepository;


    // -----------------------------------------------------------
    // Standard CRUD Methods
    // -----------------------------------------------------------

    /**
     * Processes a payment.
     *
     * @param payment the payment entity containing payment details.
     * @return the payment entity indicating success or failure.
     */
    public Payment processPayment(Payment payment) {
        // Generate a unique transaction ID using UUID for this scenario
        // But this should actually be generated dynamically from a payment gateway
        String transactionId = "TX-" + UUID.randomUUID().toString();
        payment.setTransactionId(transactionId);
        // Assume a payment is always successful in this scenario
        payment.setSuccess(true);

        // Save the payment with the generated transaction ID
        return paymentRepository.save(payment);
    }

    /**
     * Retrieve all payments.
     *
     * @return list of all payments.
     */
    public List<Payment> getAllPayments() {
        return paymentRepository.findAll();
    }

    /**
     * Retrieve a payment by ID.
     *
     * @param id the ID of the payment to retrieve.
     * @return the payment if found, Optional.empty() otherwise.
     */
    public Optional<Payment> getPaymentById(Long id) {
        return paymentRepository.findById(id);
    }

    /**
     * Update an existing payment.
     *
     * @param id the ID of the payment to update.
     * @param updatedPayment the payment object with updated data.
     * @return the updated payment if found and updated, Optional.empty() otherwise.
     */
    public Optional<Payment> updatePayment(Long id, Payment updatedPayment) {
        if (!paymentRepository.existsById(id)) {
            return Optional.empty();
        }
        updatedPayment.setId(id);
        return Optional.of(paymentRepository.save(updatedPayment));
    }

    /**
     * Delete a payment by ID.
     *
     * @param id the ID of the payment to delete.
     */
    public void cancelPayment(Long id) {
        paymentRepository.deleteById(id);
    }

    // -----------------------------------------------------------
    // Custom Methods
    // -----------------------------------------------------------

    /**
     * Retrieve payments by user ID.
     *
     * @param userId the ID of the user whose payments to retrieve.
     * @return list of payments for the specified user.
     */
    public List<Payment> getPaymentsByUser(int userId) {
        return paymentRepository.findByUserId(userId);
    }

    /**
     * Retrieve payments with amount greater than the specified value.
     *
     * @param amount the amount threshold.
     * @return list of payments with amount greater than the specified value.
     */
    public List<Payment> getPaymentsByAmountGreaterThan(BigDecimal amount) {
        return paymentRepository.findByAmountGreaterThan(amount);
    }

}

File: frontend-vue/src/services/paymentService.ts
import axios from "axios";

// Define the API base URL
const API_BASE_URL = "http://localhost:8084/api/payments";

// Define TypeScript interfaces to match the Payment entity
interface Payment {
    id?: number;
    userId: number;
    bookingId: number;
    amount: number;
    paymentMethod: string;
    transactionId?: string;
    success?: boolean;
}

// Create a Payment Service
class PaymentService {
    // Process a payment
    async processPayment(payment: Payment): Promise<Payment> {
        try {
            const response = await axios.post<Payment>(
                `${API_BASE_URL}/process`,
                payment
            );
            return response.data;
        } catch (error) {
            console.error("Error processing payment:", error);
            throw new Error("Failed to process payment");
        }
    }

    // Retrieve all payments
    async getAllPayments(): Promise<Payment[]> {
        try {
            const response = await axios.get<Payment[]>(API_BASE_URL);
            return response.data;
        } catch (error) {
            console.error("Error fetching payments:", error);
            throw new Error("Failed to fetch payments");
        }
    }

    // Retrieve a payment by ID
    async getPaymentById(id: number): Promise<Payment> {
        try {
            const response = await axios.get<Payment>(`${API_BASE_URL}/${id}`);
            return response.data;
        } catch (error) {
            console.error(`Error fetching payment with ID ${id}:`, error);
            throw new Error("Failed to fetch payment");
        }
    }

    // Update an existing payment
    async updatePayment(id: number, payment: Payment): Promise<Payment> {
        try {
            const response = await axios.put<Payment>(
                `${API_BASE_URL}/${id}`,
                payment
            );
            return response.data;
        } catch (error) {
            console.error(`Error updating payment with ID ${id}:`, error);
            throw new Error("Failed to update payment");
        }
    }

    // Delete a payment
    async cancelPayment(id: number): Promise<void> {
        try {
            await axios.delete(`${API_BASE_URL}/${id}`);
        } catch (error) {
            console.error(`Error canceling payment with ID ${id}:`, error);
            throw new Error("Failed to cancel payment");
        }
    }

    // Retrieve payments by user ID
    async getPaymentsByUser(userId: number): Promise<Payment[]> {
        try {
            const response = await axios.get<Payment[]>(
                `${API_BASE_URL}/user/${userId}`
            );
            console.log(`Fetching payments for user ID ${userId}:`);
            return response.data;
        } catch (error: any) {
            console.error(
                `Error fetching payments for user ID ${userId}:`,
                error.response || error.message
            );
            throw new Error("Failed to fetch payments by user");
        }
    }

    // Retrieve payments with amount greater than a specified value
    async getPaymentsByAmountGreaterThan(amount: number): Promise<Payment[]> {
        try {
            const response = await axios.get<Payment[]>(
                `${API_BASE_URL}/amountGreaterThan/${amount}`
            );
            return response.data;
        } catch (error) {
            console.error(
                `Error fetching payments greater than ${amount}:`,
                error
            );
            throw new Error("Failed to fetch payments by amount");
        }
    }
}

export default new PaymentService();
